tony-core/src/main/java/com/linkedin/tony/tensorflow/TonySession.java

```diff
@@ -306,8 +310,18 @@ public void updateSessionStatus() {
     }
 
     if (failureCount > 0) {
-      setFinalStatus(FinalApplicationStatus.FAILED,
-          "At least one job task exited with non-zero status, failedCnt=" + failureCount);
+      // if `tony.application.fail-on-worker-failure-enabled` is true and there is at least one failed worker,
+      // or if all worker jobs failed, we finally return "FAILED" otherwise, we return "SUCCEEDED" with some
+      // failure workers
```

Comment:
I would also mention that the reason why we try to catch the failure again here is because when the AM killed the worker task, the `onTaskCompleted()` call won't be called to trigger the short-circuit